### ISSUE TITLE
Train YOLOv11/v8 on Intel Arc Discrete GPU

### ISSUE THREAD

**Issue by ramesh-dev-code** [ultralytics/ultralytics#19821]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

What are the required changes to be made in ultralytics repository to train the YOLOv8/v11 model on Intel Arc Discrete GPUs?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ramesh-dev-code, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/), where you can find detailed information about YOLO usage and configuration. Specifically, the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) guides might be helpful for your exploration.

If this is a ❓ Question regarding custom hardware compatibility, please provide detailed information about your setup, such as the Intel Arc GPU model, your operating system, and any relevant logs or error messages. This helps us better understand your environment.

For 🐛 Bug Reports, we kindly ask for a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to assist in debugging.

To stay connected with the Ultralytics community, you can:
- Join our [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time discussions.
- Dive into our [Discourse](https://community.ultralytics.com/) for in-depth conversations.
- Engage with peers on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Ensure you're using the latest version of the `ultralytics` package and dependencies. Update everything in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) by running:

```bash
pip install -U ultralytics
```

## Environments

YOLO runs on a variety of environments with pre-installed dependencies. Here are some verified setups you can try:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response, but rest assured, an Ultralytics engineer will review your issue and provide further assistance soon. Thank you for being a part of the Ultralytics community! 😊

**Comment by ramesh-dev-code** [comment]
Install Intel oneAPI and Intel-optimized PyTorch along with a few changes in ultralytics repository to select the Intel GPU for model training. Refer to this [repository](https://github.com/ramesh-dev-code/polyp-detection-yolov11n) for the BKM to train YOLOv11n model on a custom dataset using Intel Arc A770 discrete GPU.

**Comment by glenn-jocher** [comment]
@ramesh-dev-code to train YOLO11/YOLOv8 on Intel Arc GPUs, you'll need **Intel's optimized PyTorch** with oneAPI support. While Ultralytics YOLO doesn't require code modifications for Intel Arc GPUs, you should:

1. Install [Intel Extension for PyTorch](https://intel.github.io/intel-extension-for-pytorch/)
2. Use standard training commands with `device='gpu'` after configuring your environment

For inference optimization, see our [OpenVINO integration guide](https://docs.ultralytics.com/integrations/openvino/) showing 8.5ms inference times on Arc GPUs. Community support is available via [GitHub Discussions](https://github.com/ultralytics/ultralytics/discussions).

**Comment by ramesh-dev-code** [comment]
@glenn-jocher Thanks for your suggestion. Since v2.5, Intel GPU is natively supported by PyTorch for model training with a few changes in ultralytics repository and there is no need to install Intel Extension for PyTorch. Please refer to the repository mentioned in my previous comment for the BKM.

**Comment by glenn-jocher** [comment]
@ramesh-dev-code thanks for highlighting PyTorch's native Intel GPU support! You're correct that Ultralytics YOLO works with Intel Arc GPUs via standard PyTorch implementations. For optimization best practices, we recommend our [OpenVINO integration guide](https://docs.ultralytics.com/integrations/openvino/) which achieves 8.5ms inference times on Arc GPUs. Community insights are always welcome in [GitHub Discussions](https://github.com/ultralytics/ultralytics/discussions). 🚀

**Comment by ZailiWang** [comment]
> [@ramesh-dev-code](https://github.com/ramesh-dev-code) to train YOLO11/YOLOv8 on Intel Arc GPUs, you'll need **Intel's optimized PyTorch** with oneAPI support. While Ultralytics YOLO doesn't require code modifications for Intel Arc GPUs, you should:
> 
> 1. Install [Intel Extension for PyTorch](https://intel.github.io/intel-extension-for-pytorch/)
> 2. Use standard training commands with `device='gpu'` after configuring your environment
> 
> For inference optimization, see our [OpenVINO integration guide](https://docs.ultralytics.com/integrations/openvino/) showing 8.5ms inference times on Arc GPUs. Community support is available via [GitHub Discussions](https://github.com/ultralytics/ultralytics/discussions).

Hi, looks like some changes are still required for the parts hard-coded with cuda to enable Yolo v8/v11 on Intel Arc devices. Please refer to my comment in the linked issue.

**Comment by glenn-jocher** [comment]
@ZailiWang thanks for highlighting potential CUDA dependencies in the codebase. We welcome community contributions via [Ultralytics YOLO GitHub Issues](https://github.com/ultralytics/ultralytics/issues) to improve Intel Arc compatibility. For training optimizations, feel free to add your Best Known Methods (BKMs) to our [OpenVINO integration documentation](https://docs.ultralytics.com/integrations/openvino/) through a pull request.

**Comment by jiekechoo** [comment]
@glenn-jocher Hi, I faced an issue with training YOLO in Intel ARC A770 dGPU, and how to improve compatibility with Intel arch GPU ? Any suggestion is welcome. 

Thanks.
Jieke Choo

**Comment by jiekechoo** [comment]
@ramesh-dev-code Hi, i refered your [repo](https://github.com/ramesh-dev-code/polyp-detection-yolov11n) with my training task. It's worked. But I got some errors in the result of YOLO. https://github.com/intel/intel-extension-for-pytorch/issues/799

> Install Intel oneAPI and Intel-optimized PyTorch along with a few changes in ultralytics repository to select the Intel GPU for model training. Refer to this [repository](https://github.com/ramesh-dev-code/polyp-detection-yolov11n) for the BKM to train YOLOv11n model on a custom dataset using Intel Arc A770 discrete GPU.

**Comment by ramesh-dev-code** [comment]
> [@ramesh-dev-code](https://github.com/ramesh-dev-code) Hi, i refered your [repo](https://github.com/ramesh-dev-code/polyp-detection-yolov11n) with my training task. It's worked. But I got some errors in the result of YOLO. [intel/intel-extension-for-pytorch#799](https://github.com/intel/intel-extension-for-pytorch/issues/799)
> 
> > Install Intel oneAPI and Intel-optimized PyTorch along with a few changes in ultralytics repository to select the Intel GPU for model training. Refer to this [repository](https://github.com/ramesh-dev-code/polyp-detection-yolov11n) for the BKM to train YOLOv11n model on a custom dataset using Intel Arc A770 discrete GPU.

I noticed that you closed this issue on the other link. Let me know if you are still facing any issues after modifying the below changes in [torch_utils.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/torch_utils.py)

```
def select_device(device="", batch=0, newline=False, verbose=True):
   return torch.device("xpu")
```

**Comment by azhuvath** [comment]
https://github.com/issues/mentioned?issue=ultralytics%7Cultralytics%7C20964

**Comment by glenn-jocher** [comment]
@azhuvath thanks for the reference! Could you provide more context about the specific issue you're encountering with Intel Arc GPU training? If you're facing similar device selection challenges, the community solutions discussed above regarding `torch.device("xpu")` modifications may be relevant to your case.

**Comment by azhuvath** [comment]
> [@azhuvath](https://github.com/azhuvath) thanks for the reference! Could you provide more context about the specific issue you're encountering with Intel Arc GPU training? If you're facing similar device selection challenges, the community solutions discussed above regarding `torch.device("xpu")` modifications may be relevant to your case.

@glenn-jocher The only pending issue on my side is multi card training. At present it works fine on my side with single card. At present I don't have a system with multi card. Will analyze what needs to be changed to make it working.

**Comment by glenn-jocher** [comment]
@azhuvath great to hear single card training is working well! For multi-card Intel Arc training, you'll likely need to adapt the device selection logic in `select_device()` to handle multiple `xpu` devices similar to how it currently handles CUDA devices. The function already includes multi-GPU batch size validation that should apply to Intel Arc as well - feel free to share your findings when you test with multi-card setups.
